### ISSUE TITLE
Facebook auth with already existing email

### ISSUE THREAD

**Issue by bertocq** [consul/consul#2038]
If an user tries to register/login with facebook and his account has the same email as an existing user we'll get a internal error:

`PG::UniqueViolation: ERROR: duplicate key value violates unique constraint "index_users_on_email"
DETAIL: Key (email)=(user@email.com) already exists.`

at app/controllers/users/omniauth_callbacks_controller.rb" line 46 in save_user

We need to reproduce this error locally and find a solution

## Notes

There's a similar (maybe the same?) bug, which can be reproduced on a Rails console:

```ruby
user = User.create(email: "unique@consul.dev", terms_of_service: true, username: "unique", password: "12345678")
user.hide
User.create(email: "unique@consul.dev", terms_of_service: true, username: "unique", password: "12345678")
```

With that code, an exception occurs:
```
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_users_on_email"
DETAIL:  Key (email)=(unique@consul.dev) already exists.
```

**Comment by aitbw** [comment]
I'm gonna go with this one.

**Comment by aitbw** [comment]
**UPDATE**: As today (18/12/2017), this issue has been reproduced successfully on a local (development) environment after following the subsequent steps:

* Sign up with a social network account
* Confirm the account if necessary
* `bundle exec rails c` on your terminal
* Search the account with `@user = User.find_by(email: 'your_account@email.com')`
* `@user.destroy!`
* Finally, try to sign up again with the previously used email
* Exception is raised

Obviously, this scenario is only available to CONSUL developers/sysadmins/DBAs. An user can delete his/her account through `http://localhost:3000/account`, clicking the `Erase my account` link and that account's email will be available for a new account.

The other scenario I can think of could be of a blocked user, which is soft deleted and not taken into consideration when creating a new account, causing the exception to raise.

**Comment by aitbw** [comment]
# **Update** (21/12/2017)

## TL;DR version

Exception is now avoided completely when registering with a regular CONSUL account or through Facebook. Signing through Twitter/Google+ still not working. Changes can be found [here](https://github.com/wairbut-m2c/consul/tree/aperez-duplicated-email-bug)

## Long version

### The issue

An `ActiveRecord::RecordNotUnique` exception is raised when an email belonging to a hidden (soft-deleted) user is used to register a new CONSUL account. The expected behavior is to render a message to the end-user indicating that the email address entered to create a new account is blocked and can't be used anymore

### The underlying reason

Emails belonging to hidden (soft-deleted) users are not taken into consideration when validating a record at model level since the default (User) model scope is `SELECT * FROM users WHERE hidden_at IS NOT NULL`, causing the aforementioned exception to raise when inserting said record into the database —whilst an index exists at DB level to guarantee record uniqueness, a scope to ignore hidden users does not

### Analyzing source code on `master`

[This line](https://github.com/wairbut-m2c/consul/blob/d534b02b3c2ee186e0c690d31d7dca4eceb9c176/app/controllers/users/omniauth_callbacks_controller.rb#L44) found under `app/controllers/users/omniauth_callbacks_controller.rb` was, apparently, one of the reasons the exception was being raised —that's because `@user.save` will always be executed, as stated by DeMorgan's boolean laws.

![photo_2017-12-21_18-55-38](https://user-images.githubusercontent.com/9470839/34278354-a6c90dbe-e680-11e7-8063-47f968c03347.jpg)

Since `@user.save_requiring_finish_signup` method is never called, the exception is never handled properly, hence the `500 Internal Server` error

### Alternatives to the issue

* Partial indexes (Mentioned [here](https://github.com/rubysherpas/paranoia#about-indexes), on rubysherpas/paranoia#271 and rubysherpas/paranoia#319):

Seemed like a solution where the users' email uniqueness database index is modified to look like this:

`add_index "users", ["email"], name: "index_users_on_email", unique: true, where: "hidden_at IS NOT NULL", using: :btree`

But it does not work as intended, since it allows 2 users (one hidden and one **not** hidden) to use the same email address

`add_index "users", ["email"], name: "index_users_on_email", unique: true, where: "hidden_at IS NULL", using: :btree` won't work as well since it overwrites the previous user (allowing the "new" user to login successfully) with the conflicting email address and that's not the expected behavior

* Validation at model level with conditions for uniqueness (Mentioned on rubysherpas/paranoia#333):

This looked promising, since only one (1) SLOC would be added to the User model which was the following:

`validates :email, uniqueness: { conditions: -> { with_hidden } }`, but apparently it does nothing. And `validates :email, uniqueness: { conditions: -> { User.with_hidden } }` is even worse since it considers every email address already in use

### Partial working solution (for Facebook OAuth and regular CONSUL accounts)

* For Facebook OAuth: The `save_requiring_finish_signup` class method was removed from the User model, the private `save_user` method belonging to `Users::OmniauthCallbacksController` was rewritten to check if an user (including soft-deleted records) with a certain email address exists. If `true`, the user is redirected to the `new_user_registration` url with a warning that reads `This email address has been blocked`

* For regular CONSUL accounts: The `ActiveRecord::RecordNotUnique` exception is caught and responds the same way through registration with Facebook OAuth. This change does not involves a query to the database like the former solution

### Account creation with Twitter OAuth and Google OAuth2:

* Twitter OAuth: The `ActiveRecord::RecordNotUnique` exception is raised as soon as the user clicks the confirmation link he/she received if a hidden (soft-deleted) email address is used

* Google OAuth2: Raises an exception regarding exceeded cookies max size —apparently, it has to do with the way conflicting email addresses are dealt with when signing up through Facebook OAuth

**Comment by aitbw** [comment]
# UPDATE (26/12/17)

Account creation through **Google OAuth2** now works as expected. Latest commit on [this branch](https://github.com/wairbut-m2c/consul/commits/aperez-duplicated-email-bug) amended accordingly to include the new (working) changes.

**Twitter OAuth** still not working as intended.

**Comment by aitbw** [comment]
# Update (27/12/17)

Account creation through **Twitter OAuth** now working as intended after activating the "Request email addresses from users" option on the "Twitter Application Management" dashboard, under the "Permissions" tab. Translations in both English and Spanish were also added.

![screenshot-2017-12-27 twitter application management](https://user-images.githubusercontent.com/9470839/34393297-caf7b2d4-eb27-11e7-880e-ee7c9a97dd89.png)

Both Consumer Key and Consumer Secret **must be** regenerated in order for this to work.

There's a caveat: There are four (4) failing specs (all OAuth-related), as you can see [here](https://travis-ci.org/wairbut-m2c/consul/jobs/321998857)

![screenshot-2017-12-27 job 414 2 - wairbut-m2c consul - travis ci](https://user-images.githubusercontent.com/9470839/34393392-719308f0-eb28-11e7-8021-abb73b3d2b0b.png)

The first three specs are pretty easy to get them green again but the last one is kinda tricky since the confirmation email is not being delivered —I assume this has to do with the way email addresses are being handled now.

**Comment by javierm** [comment]
Reopening as it's got the "Bug" label.

**Comment by lucialuzuriaga** [comment]
Closing as fixed.
